File: wsb_signals/extract.py
"""Ticker extractor — candidate regex + stoplist (+ optional whitelist + ambiguous gate).

Ported from scripts/e2e_top5.py and made configurable. The extractor is the
foundation of every empirical signal, so precision matters:
  - a $-cashtag ($NVDA) is high-confidence and overrides the stoplist AND the
    ambiguous gate;
  - bare uppercase tokens are dropped if they hit the stoplist or (when a
    whitelist is loaded) aren't known listed symbols;
  - AMBIGUOUS tokens (real tickers that are also common words, e.g. DRAM) are
    accepted only when the surrounding text carries TRADING CONTEXT — otherwise
    "DRAM" the memory chip would be miscounted as the Roundhill Memory ETF.
Whitelist wiring + a precision/recall eval are Phase 0.4 (architecture §2.2).
"""
from __future__ import annotations

import re
from pathlib import Path

# Optional $-cashtag + 1–5 uppercase letters. Single-letter tokens are matched only so
# classify() can ACCEPT them when $-prefixed ($F, $T) and DROP them when bare — see
# architecture.md §2.2 ("bare 1-char tokens only via an explicit $-cashtag").
DEFAULT_REGEX = r"(?<![A-Za-z0-9])(\$)?([A-Z]{1,5})(?![A-Za-z0-9])"

# Options / position vocabulary. A token co-occurring with any of these (or a $) sits in
# trading context — the cheap proxy used to (a) admit AMBIGUOUS word-tickers here and
# (b) measure the co-occurrence rate in `wsb eval-extractor` (architecture §2.2).
TRADING_WORDS = frozenset({
    "call", "calls", "put", "puts", "strike", "strikes", "expiry", "expiration", "leaps",
    "contracts", "contract", "premium", "otm", "itm", "atm", "theta", "gamma", "delta", "vega",
    "long", "short", "shares", "bought", "sold", "buy", "sell", "position", "earnings",
    "squeeze", "bullish", "bearish", "moon", "yolo",
})

_CTX_WORD = re.compile(r"[a-z]+")


def has_trading_context(text: str | None) -> bool:
    """True if the text carries options/position language (or a $) — the §2.2 co-occurrence proxy."""
    if not text:
        return False
    if "$" in text:
        return True
    return any(t in TRADING_WORDS for t in _CTX_WORD.findall(text.lower()))


class TickerExtractor:
    def __init__(
        self,
        stoplist: set[str],
        *,
        regex: str = DEFAULT_REGEX,
        whitelist: set[str] | None = None,
        ambiguous: set[str] | None = None,
    ):
        self.stop = stoplist
        self.re = re.compile(regex)
        self.whitelist = whitelist  # None = accept any non-stop candidate (no whitelist yet)
        self.ambiguous = ambiguous or set()  # word-tickers gated on trading context

    @classmethod
    def from_files(
        cls,
        stoplist_path: Path,
        *,
        regex: str = DEFAULT_REGEX,
        whitelist_path: Path | None = None,
        ambiguous_path: Path | None = None,
    ) -> "TickerExtractor":
        stop = _load_wordset(stoplist_path)
        wl = _load_wordset(whitelist_path) if whitelist_path else None
        amb = _load_wordset(ambiguous_path) if ambiguous_path else None
        return cls(stop, regex=regex, whitelist=wl, ambiguous=amb)

    @classmethod
    def cashtag_only(cls, stoplist_path: Path, *, regex: str = DEFAULT_REGEX) -> "TickerExtractor":
        """Fail-closed fallback when the whitelist is configured but missing: an EMPTY whitelist
        rejects every bare token (`not_listed`), so only high-confidence $-cashtags are accepted."""
        return cls(_load_wordset(stoplist_path), regex=regex, whitelist=set())

    # Decision codes from classify(); accepted = cashtag / whitelist_ok / open_ok.
    ACCEPT = ("cashtag", "whitelist_ok", "open_ok")

    def classify(self, text: str | None) -> list[tuple[str, str]]:
        """Per candidate token, return (symbol, decision). Shared by extract() and the eval.

        Decisions: cashtag | too_short | stop | not_listed | ambig_no_context | whitelist_ok | open_ok.
        An ambiguous word-ticker without a $-cashtag is admitted only when the text as a whole
        carries trading context (computed once per text, not per token).
        """
        out: list[tuple[str, str]] = []
        # Only pay for the context scan when this text actually has an ambiguous candidate.
        ctx: bool | None = None
        for m in self.re.finditer(text or ""):
            cash, sym = m.group(1), m.group(2)
            if cash:
                d = "cashtag"        # $-cashtag: high confidence, bypasses stop / whitelist / ambiguous
            elif len(sym) < 2:
                d = "too_short"      # bare single letter (e.g. "F") — too noisy without a cashtag
            elif sym in self.stop:
                d = "stop"           # common word / usually-not-a-ticker (even if also listed)
            elif self.whitelist is not None and sym not in self.whitelist:
                d = "not_listed"     # not a real listed symbol → drop
            elif sym in self.ambiguous:
                if ctx is None:
                    ctx = has_trading_context(text)
                d = ("whitelist_ok" if self.whitelist is not None else "open_ok") if ctx else "ambig_no_context"
            else:
                d = "whitelist_ok" if self.whitelist is not None else "open_ok"
            out.append((sym, d))
        return out

    def extract(self, text: str | None) -> list[str]:
        return [sym for sym, d in self.classify(text) if d in self.ACCEPT]


def _load_wordset(path: Path) -> set[str]:
    words: set[str] = set()
    for line in Path(path).read_text().splitlines():
        line = line.split("#", 1)[0].strip()
        if line:
            words.update(line.split())
    return words

File: packages/worker/src/extract.ts
/**
 * Ticker extractor — TS port of the frozen v0.0.1 `wsb_signals/extract.py` (architecture §2.2).
 *
 * Parity contract: gated on fixtures/extract_classify.json + fixtures/wordset_loader.json
 * (v2-porting-spec.md §3). Decision precedence is exact:
 *   cashtag > too_short > stop > not_listed > ambiguous-gate > accept.
 *
 * Cross-language landmines handled here:
 *  - the candidate regex stays ASCII (no `u` flag) and MUST be global for `matchAll`;
 *  - `match[1]` is `undefined` (not Python's `None`) when the optional `$` group is absent — the
 *    `if (cash)` truthiness check is equivalent;
 *  - the wordset loader mirrors Python `str.split()`: split on ANY whitespace run, drop empties.
 */

/** Optional `$`-cashtag + 1–5 uppercase letters, bounded by non-alnum. Group 1 = `$`, group 2 = symbol.
 *  Global flag is required by `matchAll`; `matchAll` clones the regex, so sharing this is lastIndex-safe. */
export const DEFAULT_REGEX = /(?<![A-Za-z0-9])(\$)?([A-Z]{1,5})(?![A-Za-z0-9])/g

/** Options / position vocabulary — the §2.2 trading-context proxy (admits ambiguous word-tickers). */
export const TRADING_WORDS: ReadonlySet<string> = new Set([
  'call', 'calls', 'put', 'puts', 'strike', 'strikes', 'expiry', 'expiration', 'leaps',
  'contracts', 'contract', 'premium', 'otm', 'itm', 'atm', 'theta', 'gamma', 'delta', 'vega',
  'long', 'short', 'shares', 'bought', 'sold', 'buy', 'sell', 'position', 'earnings',
  'squeeze', 'bullish', 'bearish', 'moon', 'yolo',
])

const CTX_WORD = /[a-z]+/g

/** True if the text carries options/position language (or a `$`) — the §2.2 co-occurrence proxy. */
export function hasTradingContext(text: string | null | undefined): boolean {
  if (!text) return false
  if (text.includes('$')) return true
  for (const m of text.toLowerCase().matchAll(CTX_WORD)) {
    if (TRADING_WORDS.has(m[0])) return true
  }
  return false
}

export type Decision =
  | 'cashtag' | 'too_short' | 'stop' | 'not_listed' | 'ambig_no_context' | 'whitelist_ok' | 'open_ok'

/** accepted = cashtag / whitelist_ok / open_ok (mirrors `TickerExtractor.ACCEPT`). */
const ACCEPT: ReadonlySet<Decision> = new Set<Decision>(['cashtag', 'whitelist_ok', 'open_ok'])

export interface ExtractorOptions {
  regex?: RegExp
  whitelist?: ReadonlySet<string> | null // null = accept any non-stop candidate (no whitelist loaded)
  ambiguous?: ReadonlySet<string> | null
}

export class TickerExtractor {
  private readonly stop: ReadonlySet<string>
  private readonly re: RegExp
  private readonly whitelist: ReadonlySet<string> | null
  private readonly ambiguous: ReadonlySet<string>

  constructor(stop: ReadonlySet<string>, opts: ExtractorOptions = {}) {
    this.stop = stop
    const re = opts.regex ?? DEFAULT_REGEX
    // `matchAll` throws on a non-global regex; ensure `g` so a config-supplied pattern still works.
    this.re = re.global ? re : new RegExp(re.source, `${re.flags}g`)
    this.whitelist = opts.whitelist ?? null
    this.ambiguous = opts.ambiguous ?? new Set()
  }

  /** Fail-closed fallback: an EMPTY whitelist rejects every bare token, so only `$`-cashtags pass. */
  static cashtagOnly(stop: ReadonlySet<string>, opts: { regex?: RegExp } = {}): TickerExtractor {
    return new TickerExtractor(stop, { regex: opts.regex, whitelist: new Set() })
  }

  /** Per candidate token, return [symbol, decision] in left-to-right match order. */
  classify(text: string | null | undefined): Array<[string, Decision]> {
    const out: Array<[string, Decision]> = []
    let ctx: boolean | null = null // compute trading context at most once per text
    for (const m of (text ?? '').matchAll(this.re)) {
      const cash = m[1]
      const sym = m[2] as string
      let d: Decision
      if (cash) {
        d = 'cashtag' // $-cashtag: high confidence, bypasses stop / whitelist / ambiguous
      } else if (sym.length < 2) {
        d = 'too_short' // bare single letter — too noisy without a cashtag
      } else if (this.stop.has(sym)) {
        d = 'stop'
      } else if (this.whitelist !== null && !this.whitelist.has(sym)) {
        d = 'not_listed'
      } else if (this.ambiguous.has(sym)) {
        if (ctx === null) ctx = hasTradingContext(text)
        d = ctx ? (this.whitelist !== null ? 'whitelist_ok' : 'open_ok') : 'ambig_no_context'
      } else {
        d = this.whitelist !== null ? 'whitelist_ok' : 'open_ok'
      }
      out.push([sym, d])
    }
    return out
  }

  extract(text: string | null | undefined): string[] {
    return this.classify(text).filter(([, d]) => ACCEPT.has(d)).map(([sym]) => sym)
  }
}

/**
 * Parse a stoplist/whitelist/ambiguous file body into a set — port of `_load_wordset`.
 * Mirrors Python exactly: take text before the first `#`, strip, split on ANY whitespace run
 * (dropping empties), dedup. Pass the file CONTENTS (callers own fs); splits on LF/CRLF/CR.
 */
export function loadWordset(content: string): Set<string> {
  const words = new Set<string>()
  for (const rawLine of content.split(/\r\n|\r|\n/)) {
    const line = (rawLine.split('#', 1)[0] ?? '').trim()
    if (!line) continue
    for (const w of line.split(/\s+/)) {
      if (w) words.add(w)
    }
  }
  return words
}
